internal/k6runner/version/repository.go

```diff
@@ -0,0 +1,239 @@
+package version
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/rs/zerolog"
+)
+
+// Repository points to a collection of k6 binaries.
+type Repository struct {
+	// Root points to a folder in the local filesystem that will be scanned for k6 binaries.
+	// All executable files in said folder should be k6 binaries, as they will all be executed with `--version` to map
+	// their actual versions.
+	Root string
+	// Override is the path to a specific k6 binary. If set, all Repository operations will return this path.
+	Override string
+
+	// Logger.
+	Logger zerolog.Logger
+
+	mtx     sync.Mutex
+	entries []Entry
+}
+
+const binaryMustContain = "k6"
+
+// Entry stores the path to a k6 binary, and the parsed semantic version that binary returns.
+type Entry struct {
+	Path    string
+	Version *semver.Version
+}
+
+type k6Version struct {
+	Commit    string `json:"commit"`
+	GoArch    string `json:"go_arch"`
+	GoOs      string `json:"go_os"`
+	GoVersion string `json:"go_version"`
+	Version   string `json:"version"`
+}
+
+// ErrUnsatisfiable is returned when no binary matching a constraint is found.
+var ErrUnsatisfiable = errors.New("no compatible k6 version found")
+
+// BinaryFor walks the repository and returns the path to the binary with the highest version that matches
+// constraintStr. If no binary matching the constraint is found, ErrUnsatisfiable is returned.
+// BinaryFor scans the root folder if needed.
+func (r *Repository) BinaryFor(constraintStr string) (string, error) {
```

Comment:
To understand better the usage of this method, how do we expect to use this? Currently AFAIK the k6 runner implementation accepts a static "k6 path". I guess that will be changed so the path is resolved as part of the [`Run()`](https://github.com/grafana/synthetic-monitoring-agent/blob/191471cc59055fc665567118ee66dc3481831f0c/internal/k6runner/k6runner.go#L94) method? And therefore _run_ will be extended (possible [`Settings`](https://github.com/grafana/synthetic-monitoring-agent/blob/191471cc59055fc665567118ee66dc3481831f0c/internal/k6runner/k6runner.go#L29) to accept the specified k6 channel?